[Merged by Bors] - sync2: ATX integration

ivan4th opened the pull request:
## Motivation

We need more efficient ATX sync.

## Description

This adds set reconciliation for ATXs.
There are per-epoch syncers, with lower FPTree depth (16 by default)
used for older epochs and greater FPTree depth (21 by default) used
for current epoch.
Both active syncv2 and passive (server-only) syncv2 are disabled by
default. It is possible to enable syncv2 in server-only or
full (active) mode.

## Test Plan

Test on testnet and mainnet nodes

## TODO

- [ ] Explain motivation or link existing issue(s)
- [ ] Test changes and document test plan
- [ ] Update documentation as needed
- [ ] Update [changelog](../CHANGELOG.md) as needed

ivan4th committed:
sql: add Database.Connection/WithConnection, interface cleanup

This adds a possibility to take a connection from the pool to use it
via the Executor interface, and return it later when it's no longer
needed. This avoids connection pool overhead in cases when a lot of
quries need to be made, but the use of read transactions is not
needed.

Using read transactions instead of simple connections has the side
effect of blocking WAL checkpoints.

ivan4th committed:
sync2: dbset: use single connection for each sync session

Using single connection for multiple SQL queries which are executed
during sync avoids noticeable overhead due to SQLite connection pool
delays.

Also, this change fixes memory overuse in DBSet. When initializing
DBSet from a database table, there's no need to use an FPTree with big
preallocated pool for the new entries that are added during recent sync.

ivan4th committed:
sync2: multipeer: fix edge cases

Split sync could become blocked when there were slow peers. Their
subranges are assigned to other peers, and there were bugs causing
indefinite blocking and panics in these cases. Moreover, after other
peers managed to sync the slow peers' subranges ahead of them, we need
to interrupt syncing against the slow peers as it's no longer needed.

In multipeer sync, when every peer has failed to sync, e.g. due to
temporary connection interruption, we don't need to wait for the full
sync interval, using shorter wait time between retries.

ivan4th committed:
sync2: ATX integration

This adds set reconciliation for ATXs.
There are per-epoch syncers, with lower FPTree depth (16 by default)
used for older epochs and greater FPTree depth (21 by default) used
for current epoch.
Both active syncv2 and passive (server-only) syncv2 are disabled by
default. It is possible to enable syncv2 in server-only or
full (active) mode.

fasmat commented on config/mainnet.go:
```diff
@@ -77,6 +78,14 @@ func MainnetConfig() Config {
 	hare4conf := hare4.DefaultConfig()
 	hare4conf.Enable = false
+
+	oldAtxSyncCfg := sync2.DefaultConfig()
+	oldAtxSyncCfg.MultiPeerReconcilerConfig.SyncInterval = time.Hour
+	oldAtxSyncCfg.MaxDepth = 16
+	newAtxSyncCfg := sync2.DefaultConfig()
+	newAtxSyncCfg.MaxDepth = 21
+	newAtxSyncCfg.MultiPeerReconcilerConfig.SyncInterval = 5 * time.Minute
```
I don't like how this config is structured, what is the meaning of `OldAtxSyncCfg` and `NewAtxSyncCfg`? 
Is there maybe a better name for them?

Why are they part of `"main"` -> `"Syncer"` -> `"V2"`? Shouldn't they be part of the `"main"` -> `"Syncer"` config object?

Also if these are the defaults anyway, no need to overwrite them with the same values here again 🙂

fasmat replied:
EDIT: fixed spelling

ivan4th replied:
This might cause confusion as there's currently V1 sync for ATXs, too, which is also configured in the syncer config.
When V2 is used in server-only mode, both V1 and V2 syncs are used at the same time.

ivan4th replied:
The idea is to use reasonable defaults so that most users will not have to adjust sync settings.
Then at some time in the future we might want to restructure the config to remove all the v1 stuff and keep v2 sync only, moving it out of this `v2` field.

ivan4th replied:
old vs new in this context means old epochs vs the current epoch

fasmat replied:
Ah so it is actually `PrevEpochSyncConfig` and `CurrentEpochSyncConfig`? 🙂

fasmat replied:
> The idea is to use reasonable defaults so that most users will not have to adjust sync settings. Then at some time in the future we might want to restructure the config to remove all the v1 stuff and keep v2 sync only, moving it out of this `v2` field.

The reasonable defaults are already specified as part of the `syncer.DefaultConfig`, my point was that we don't need to overwrite this values again with the same values in `config.MainnetConfig` 🙂

fasmat replied:
> This might cause confusion as there's currently V1 sync for ATXs, too, which is also configured in the syncer config. When V2 is used in server-only mode, both V1 and V2 syncs are used at the same time.

But this is a technical detail that I think should not be exposed via the config, or at least in a different way. Maybe instead of `V2` we can call it something like "reconciliationSync" or similar. Especially since "V2" has negative connotations: https://en.wikipedia.org/wiki/V-2_rocket

ivan4th replied:
Actually there's a bit of confusion re current vs prev, and I was imprecise in my above comment.
When a new epoch starts, activations don't start to be published in that epoch immediately, and consequently, the more efficient (but more memory and CPU hungry per missing/extra set element) sync is to be applied to the new epoch some time later (`EpochEndFraction` in the existing sync config). That's why the "new" epoch is not necessarily the current one, and the "old" epoch is not necessarily the previous. We could probably come up with better naming but I'm unsure `current` / `prev` is the right choice here.

Otherwise agree, will fix

ivan4th replied:
Renamed to `v2` part to `reconc-sync` (`ReconcSync`).

ivan4th committed:
sync2: dbset: fix connection leak in non-loaded DBSets

ivan4th committed:
Merge branch 'sync2/dbset-conns' into sync2/atxs

ivan4th committed:
sql: revert removing Rollback method of the Migration interface

ivan4th committed:
sql: remove Database.Connection() method, keep WithConnection()

ivan4th committed:
Merge branch 'feature/long-db-conns' into sync2/dbset-conns

ivan4th committed:
sql: allow multiple connections to in-memory database

ivan4th committed:
sync2: fixup for temporary OrderedSet copies

ivan4th committed:
Merge branch 'sync2/dbset-conns' into sync2/fix-multipeer

ivan4th committed:
Merge branch 'sync2/fix-multipeer' into sync2/atxs

ivan4th committed:
sync2: use saner retry scheme for fetched ATXs

ivan4th committed:
Merge branch 'develop' into sync2/dbset-conns

ivan4th committed:
Merge branch 'sync2/dbset-conns' into sync2/atxs

ivan4th committed:
syncer: rename "v2" field to "reconcSync" in the config

ivan4th committed:
sync2: add server options and request rate limits

ivan4th committed:
Merge branch 'sync2/dbset-conns' into sync2/fix-multipeer

ivan4th committed:
Merge branch 'develop' into sync2/atxs

ivan4th committed:
sync2: fix mainnet/testnet configs

ivan4th committed:
sync2: only handle synced keys during commit

It turned out that sync interactions are happening rather quickly,
and thus it is not really practical to try and begin handling arriving
keys (e.g. ATX IDs) during sync itself. Moreover, on-the-fly key
handling was actually only used to register peer IDs for each ATX ID
to fetch the actual blob from, and that can be done in the handler's
Commit() method just as well.

ivan4th committed:
Merge branch 'develop' into sync2/fix-multipeer

ivan4th committed:
Merge branch 'develop' into sync2/atxs

ivan4th committed:
Merge branch 'sync2/fix-multipeer' into sync2/atxs

ivan4th committed:
Merge branch 'develop' into sync2/atxs

fasmat commented on sync2/atxs.go:
```diff
@@ -0,0 +1,326 @@
+				numDownloaded++
+				someSucceeded = true
+				delete(state, id)
+			case errors.Is(err, pubsub.ErrValidationReject):
+				// if the atx invalid there's no point downloading it again
+				state[id] = h.maxAttempts
```
I'm not sure this is correct; `pubsub.ErrValidationReject` does not mean the ATX is invalid, it means the handler or fetcher decided that the peer should be dropped. This can happen for a multitude of reasons: the ATX failed to decode, a dependency of the ATX failed to be fetched or the ATX is syntactically invalid.

This also means that we drop the connection to the peer we just fetched data from.

ivan4th replied:
This logic is copied over from the existing ATX syncer. The rationale is this: after this error, there's not much sense in trying to download this particular ATX. Given that reconciliation is done against multiple peers, that should not be too much of an issue. Yet, this measure guards against the peers that want to "play" the sync, trying to feed it IDs of ATXs that don't really exist / are not really valid.

fasmat commented on sync2/atxs.go:
```diff
@@ -0,0 +1,326 @@
+				state[id]++
+			}
+		}))
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
+				return err
+			}
+			if !errors.Is(err, &fetch.BatchError{}) {
+				h.logger.Debug("failed to download ATXs", zap.Error(err))
+			}
+		}
```
NIT:

```suggestion
		switch {
		case err == nil:
		case errors.Is(err, context.Canceled):
			return err
		case !errors.Is(err, &fetch.BatchError{}):
			h.logger.Debug("failed to download ATXs", zap.Error(err))
		}
```

also what if `err` is a `&fetch.BatchError{}` do we then just ignore it?

ivan4th replied:
We just log `fetch.BatchError` here b/c in this case the errors were already handled in the callbacks

ivan4th replied:
Applied the patch

fasmat commented on sync2/atxs.go:
```diff
@@ -0,0 +1,326 @@
+				zap.Int("total", total),
+				zap.Int("downloaded", numDownloaded),
+				zap.Float64("rate per sec", float64(numDownloaded)/elapsed.Seconds()))
+		}
```
NIT: `else` isn't required 🙂 

```suggestion
		if !someSucceeded {
			if batchAttemptsRemaining == 0 {
				return errors.New("failed to download ATXs: max batch retries reached")
			}
			batchAttemptsRemaining--
			h.logger.Debug("failed to download any ATXs: will retry batch",
				zap.Int("remaining", batchAttemptsRemaining),
				zap.Duration("delay", h.failedBatchDelay),
			)
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-h.clock.After(h.failedBatchDelay):
				continue
			}
		}

		batchAttemptsRemaining = h.maxBatchRetries
		elapsed := h.clock.Since(startTime)
		h.logger.Debug("fetched atxs",
			zap.Int("total", total),
			zap.Int("downloaded", numDownloaded),
			zap.Float64("rate per sec", float64(numDownloaded)/elapsed.Seconds()),
		)
```

ivan4th replied:
Fixed

fasmat commented on sync2/atxs.go:
```diff
@@ -0,0 +1,326 @@
+					zap.String("atx", id.ShortString()))
+				delete(state, id)
+				continue
+			}
```
I feel like this should be part of the `default` case in the `WithATXCallback`. This would reduce the amount of elements that need to be iterated here.

ivan4th replied:
Moved to the callback

fasmat commented on sync2/atxs.go:
```diff
@@ -0,0 +1,326 @@
+	peer p2p.Peer,
+	base rangesync.OrderedSet,
+	received rangesync.SeqResult,
+) error {
```
This method is a bit long and I think could be split up into smaller methods to aid readability 🙂

ivan4th replied:
Split `Commit()` into 3 methods

fasmat commented on sync2/atxs.go:
```diff
@@ -0,0 +1,326 @@
+	for k := range received.Seq {
+		if firstK == nil {
+			firstK = k
+		} else if firstK.Compare(k) == 0 {
+			break
+		}
```
I'm failing to understand this - why should `received.Seq` return the first element again? and why is this a break condition?

I just noticed that `SeqResult` is cyclic - that seems unintuitive to me? Iterators either always start from the first element (generally what the std library does) or are single use - i.e. cannot be iterated twice. For such a behavior I would use a ring: https://pkg.go.dev/container/ring. Apparently we already use rings in our codebase but with a different (custom) implementation: https://github.com/spacemeshos/go-spacemesh/blob/138e9bda650224c586d5d1eb232971bb6e986df0/events/reporter.go#L497-L501

What is the requirement for them being cyclic? To me it seems that this can easily cause errors like:

```go
for k := range received.Seq {
   // do something with k
} // iterates forever?
```

And these additional checks to see if we made a full cylce are necessary. Also stopping early will cause the next iteration to start at a different element as the previous - which also seems unintuitive? 🤔

ivan4th replied:
The need for cyclic iterators stems from the range representation used, see description here: https://github.com/spacemeshos/go-spacemesh/blob/develop/dev-docs/sync2-set-reconciliation.md#range-representation
Basically `RangeSetReconciler` needs `OrderedSet` it operates on to yield cyclic iterators when requesting range info, and currently no effort is made to make them non-cyclic.
On the other hand, in [Received()](https://github.com/spacemeshos/go-spacemesh/blob/9092fe701922855261d03dd0e54d4699b653445b/sync2/rangesync/interface.go#L51) method, the cyclic property is not needed. Yet, for the reason of consistency, I was trying to assume that every iterator is cyclic.

We can probably try to get rid of this cyclic property with some effort, given that strictly speaking what really needed is wraparound at the highest ID value possible not necessarily cyclicity per se, not sure whether it's worth doing right now though. Will probably try to see how hard it really is.

ivan4th replied:
Made most of the sequences non-cyclic, except for those returned by `sync2/sqlstore.IDStore` interface methods (used by `FPTree` and `DBSet` internally), as these may be trickier. Also removed most checks for cyclicity / wrapping around to the first element.

fasmat commented on sync2/atxs.go:
```diff
@@ -0,0 +1,326 @@
+			lastSynced = epoch
+		} else {
+			syncer.Start()
+		}
```
NIT: without `else` its clearer to me what's going on:

```suggestion
		if epoch > lastWaitEpoch {
			syncer.Start()
			continue
		}
		s.logger.Info("waiting for epoch to sync", zap.Uint32("epoch", epoch.Uint32()))
		if err := syncer.StartAndSync(ctx); err != nil {
			return lastSynced, fmt.Errorf("error syncing old ATXs: %w", err)
		}
		lastSynced = epoch
```

ivan4th replied:
Fixed

fasmat commented on sync2/atxs.go:
```diff
@@ -0,0 +1,326 @@
+		NewATXHandler(
+			logger, f, cfg.BatchSize, cfg.MaxAttempts,
+			cfg.MaxBatchRetries, cfg.FailedBatchDelay, nil),
+		cfg, enableActiveSync)
```
NIT:

```suggestion
	curSet := dbset.NewDBSet(db, atxsTable(epoch), 32, cfg.MaxDepth)
	handler := NewATXHandler(logger, f, cfg.BatchSize, cfg.MaxAttempts, cfg.MaxBatchRetries, cfg.FailedBatchDelay, nil)
	return NewP2PHashSync(logger, d, name, curSet, 32, f.Peers(), handler, cfg, enableActiveSync)
```

ivan4th replied:
Fixed

fasmat commented on sync2/atxs.go:
```diff
@@ -0,0 +1,326 @@
+	s.newEpoch = 0
+}
+
+func atxsTable(epoch types.EpochID) *sqlstore.SyncedTable {
+	return &sqlstore.SyncedTable{
+		TableName:       "atxs",
+		IDColumn:        "id",
+		TimestampColumn: "received",
+		Filter:          expr.MustParse("epoch = ?"),
+		Binder: func(s *sql.Statement) {
+			s.BindInt64(1, int64(epoch))
+		},
+	}
+}
```
Doesn't this already require that `id` in `atxs` is an integer? It looks to me that `sync2/sqlstore` is a lot of code to do mostly straight forward SQL queries?

ivan4th replied:
Didn't fully understand about `id` being integer requirement.
`sync2/sqlstore` provides database-backed foundation for [FPTree](https://github.com/spacemeshos/go-spacemesh/blob/138e9bda650224c586d5d1eb232971bb6e986df0/dev-docs/sync2-set-reconciliation.md#fptree-data-structure), it also contains db-bound iterator (sequence) [implementation](https://github.com/spacemeshos/go-spacemesh/blob/sync2/atxs/sync2/sqlstore/dbseq.go) that handles wraparounds of the ID range.

fasmat commented on sync2/atxs_test.go:
```diff
@@ -0,0 +1,393 @@
+func atxSeqResult(atxs []types.ATXID) rangesync.SeqResult {
+	return rangesync.SeqResult{
+		Seq: func(yield func(k rangesync.KeyBytes) bool) {
+			// Received sequence may be cyclic and the handler should stop
+			// when it sees the first key again.
+			for _, atx := range atxs {
+				if !yield(atx.Bytes()) {
+					return
+				}
+			}
```
I don't understand this comment - this sequence is not cyclic. See also my other comment about why this even needs to be cyclic at all.

ivan4th replied:
The idea is that it can be cyclic from consistency standpoint, `rangesync.Seq` being _possibly_ cyclic.
As described in other comment, I may try to get rid of this, but that takes some effort

ivan4th replied:
Got rid of cyclicity considerations here

fasmat commented on sync2/atxs_test.go:
```diff
@@ -0,0 +1,393 @@
+	defer func() {
+		cancel()
+		eg.Wait()
+	}()
```
To me this is a bid odd in the workflow. How about:

```go
ctx, cancel := context.With....
defer cancel()
var eg errgroup.Group

// some eg.Go() calls

// assertions

cancel() // explicit stop of go routines
require.NoError(t, eg.Wait())
```

I know that doesn't guarantee that `eg.Wait()` is called when the test exits, but this only happens when the test fails.

ivan4th replied:
Fixed

fasmat commented on sync2/dbset/dbset.go:
```diff
@@ -92,7 +93,7 @@ func (d *DBSet) Received() rangesync.SeqResult {
 		},
 		Error: rangesync.NoSeqError,
-	}
+	}, len(d.received)
```
Doesn't this defeat the purpose of the iterator? Why does the caller need to know how many elements are returned in advance?

ivan4th replied:
It doesn't fully defeat the purpose as using iterators is preferable to creating temporary slices which substantially increase GC pressure and thus incur not-immediately-obvious CPU costs.
Yet, in this case we only actually need to know whether the returned sequence is not empty and this can be done by checking if it contains any elements. Will fix

ivan4th replied:
Fixed, sequence length is no longer returned

fasmat commented on sync2/multipeer/interface.go:
```diff
@@ -11,38 +11,28 @@ import (
 //go:generate mockgen -typed -package=multipeer_test -destination=./mocks_test.go -source=./interface.go
```
Why Is the destination file `mocks_test.go` and not just `mocks.go`? Same applies for the `sync2/interface.go` file.

ivan4th replied:
We don't need `mocks.go` to be included in the actual binary and in this case we need to use mocks in the same test package, so we can't move it to a separate package (+subdirectory)

fasmat commented on sync2/rangesync/seq.go:
```diff
@@ -140,3 +140,20 @@ func ErrorSeqResult(err error) SeqResult {
+		},
+		Error: NoSeqError,
+	}
+}
```
See my other comments - why does this need to be cyclic?

ivan4th replied:
Removed cyclicity

fasmat commented on sync2/p2p_test.go:
```diff
@@ -21,40 +21,22 @@ import (
 	defer fh.mtx.Unlock()
-	for k := range fh.synced {
-		fh.committed[k.key] = struct{}{}
+	for k := range received.Seq {
+		fh.committed[string(k)] = struct{}{}
 	}
```
Won't this lead into an endless loop?

ivan4th replied:
"In theory" it may, will fix

ivan4th replied:
Made it so the sequence can't be cyclic here

fasmat commented on sync2/rangesync/dumbset.go:
```diff
@@ -187,7 +187,7 @@ func (ds *DumbSet) Receive(id KeyBytes) error {
 
 // Received implements the OrderedSet.
-func (ds *DumbSet) Received() SeqResult {
+func (ds *DumbSet) Received() (SeqResult, int) {
```
Same as above, why the iterator if the length is already known in advance?

ivan4th replied:
Removed length ret val

fasmat commented on syncer/syncer.go:
```diff
@@ -204,6 +247,18 @@ func NewSyncer(
+		s.dispatcher = sync2.NewDispatcher(s.logger, fetcher.(sync2.Fetcher), serverOpts)
+		hss := sync2.NewATXSyncSource(
+			s.logger, s.dispatcher, cdb.Database.(sql.StateDatabase),
+			fetcher.(sync2.Fetcher), s.cfg.ReconcSync.EnableActiveSync)
```
Why are the casts for `fetcher` and `cdb.Database` necessary? Can't `NewATXSyncSource` or `NewSyncer` be adjusted accordingly that these casts won't be necessary?

ivan4th replied:
Removed casts, had to adjust `syncer.fetcher` interface though

fasmat replied:
See my other comments about the `sync2.Fetcher` and `syncer.fetcher` interfaces

fasmat commented on syncer/syncer.go:
```diff
@@ -478,31 +553,15 @@ func (s *Syncer) syncAtx(ctx context.Context) error {
 		s.backgroundSync.epoch.Store(0)
 	}
 	if s.backgroundSync.epoch.Load() == 0 && publish.Uint32() != 0 {
+		// TODO: syncv2
```
?

ivan4th replied:
Removed stray comment

fasmat commented on syncer/syncer.go:
```diff
@@ -525,7 +584,75 @@ func (s *Syncer) syncAtx(ctx context.Context) error {
+
+func (s *Syncer) ensureMalfeasanceInSync(ctx context.Context) error {
+	// TODO: use syncv2 for malfeasance proofs:
+	// https://github.com/spacemeshos/go-spacemesh/issues/3987
```
I do not understand how this issue relates to Malfeasance sync?

ivan4th replied:
See this comment: https://github.com/spacemeshos/go-spacemesh/issues/3987#issuecomment-1866021562

fasmat replied:
I still don't understand how these two are related. The issue talks about sync issues in general and only mentions malfeasance proofs in passing. Your comment lists options for different sync methods.

1. When and how is this issue considered resolved?
2. How is this method in particular related to the issue?

ivan4th replied:
Removed the link to issue to avoid confusion. It has to deal with state sync somewhat, but is indeed more general

fasmat commented on syncer/syncer.go:
```diff
@@ -525,7 +584,75 @@ func (s *Syncer) syncAtx(ctx context.Context) error {
+			return err
+		}
+		s.logger.Info("malicious IDs synced", log.ZContext(ctx))
+		// Malfeasance proofs are synced after the actual ATXs.
+		// We set ATX synced status after both ATXs and malfeascance proofs
+		// are in sync.
+		s.setATXSynced()
```
I'm unsure about that - how is it at the moment? Aren't we considered to be ATX synced even if malfeasance sync hasn't completed yet?

ivan4th replied:
Yes, that's how it works currently: https://github.com/spacemeshos/go-spacemesh/blob/22a4f81044f9453208bdd96f900740bc77dfe4cd/syncer/syncer.go#L488

fasmat commented on syncer/syncer_test.go:
```diff
@@ -520,9 +566,72 @@ func TestSyncAtxs(t *testing.T) {
+	ts.mTicker.advanceToLayer(gLayer)
+	ts.expectMalEnsureInSync(gLayer)
+	ts.mASV2.EXPECT().EnsureSync(gomock.Any(), types.EpochID(0), types.EpochID(1)).MinTimes(1)
+	// ts.mAtxSyncer.EXPECT().Download(gomock.Any(), gLayer.GetEpoch(), gomock.Any())
```
can be removed I think

ivan4th replied:
Removed

fasmat commented on fetch/mesh_data.go:
```diff
@@ -83,18 +97,26 @@ func (f *Fetch) getHashes(
 		if err := options.limiter.Acquire(ctx, 1); err != nil {
 			pendingMetric.Add(float64(i - len(hashes)))
-			return fmt.Errorf("acquiring slot to get hash: %w", err)
+			err = fmt.Errorf("acquiring slot to get hash: %w", err)
+			for _, h := range hashes[i:] {
+				options.callback(h, err)
+			}
+			return err
```
That seems very inefficient calling possibly millions of callbacks to communicate the same error where the caller side probably doesn't even care about any error besides the first?

ivan4th replied:
It's rather likely to be a small subset of ATX IDs e.g. not downloaded due to `hs/1` request throttling, and these requests will be retried later

fasmat replied:
The point still stands - `getHashes` might be called with 5.0 Mio hashes at once - if the limit cannot be acquired because throttling is active then millions of callbacks are called.

As far as I can see this only affects this section of the code:

https://github.com/spacemeshos/go-spacemesh/blob/ad4924b998ca085bad90ebb9df2fcc16c5ab9c6c/sync2/atxs.go#L110-L129

This will ~print a debug log with the exact same error for every ATX and~ increment every element in `cs.state`. This could be handled much simpler (and arguably more efficiently) without requiring to keep track of the retries of every single hash:

Arguably this is out of the scope of this PR but this should be addressed. It makes no sense to register a hash for a peer then requesting that hash in a batch and let the fetcher again reconstruct from which peer to fetch that hash from. Error handling is also bad, because for every fetched hash an error has to be returned via callback or aggregated in a `&fetcher.BatchError{}`. Instead imo it would be much simpler to just have a (blocking) `getHash` method that fetches a hash from a given peer and returns an error if something went wrong. Then the caller can easily parallize requests and match errors to peers & hashes.

Internally the fetcher can still group requests into single batches, request those from individual peers and deserialise the batched result. This also makes it easier to figure out what when wrong if something did go wrong. Right now we have a lot of log errors of the kind `validation ran for unknown hash` and `hash missing from ongoing requests` because the fetcher fails to match requests of hashes to peers and/or callers with how it is structured at the moment.

ivan4th replied:
We should do a serious `fetcher` cleanup after we switch to `syncv2` at least for ATXs and malfeasance proofs. Right now it's probably a bit too early as we'll have to update v1 syncers that are on their way out.
Simplifying blob fetch logic that currently uses promises etc. should be one of the goals, other being removing non-streaming fetcher client and server code.

fasmat replied:
OK, I summarized all my previous concerns with the fetcher in this issue as well as liked some existing issued in it to make sure they are not forgotten: https://github.com/spacemeshos/go-spacemesh/issues/6545

fasmat commented on fetch/mesh_data.go:
```diff
@@ -83,18 +97,26 @@ func (f *Fetch) getHashes(
 		if err != nil {
 			options.limiter.Release(1)
 			pendingMetric.Add(float64(i - len(hashes)))
+			for _, h := range hashes[i:] {
+				options.callback(h, err)
+			}
```
see above

ivan4th replied:
As above

fasmat reviewed (changes requested):
Reviewing this PR took me quite some time because it builds on code that I'm not yet very familiar with. There are quite a few things that I noticed when going through the code:

We callbacks in places where they make code more complex (and inefficient) than it needs to be. Why are errors propagated using callbacks?

I'm also unsure about the DB abstraction that `syncv2` builds on. Why do we have to implement something like this ourselves? Are there no existing solutions that we can use? Is it even needed? It's another layer of abstraction for mostly simple queries as far as I can tell, it also overlaps with the existing `sql/builder` package that exists to solve mostly the same challenges and was newly implemented only a few months ago.

I'm becoming unsure about the extensive use of iterators in `syncv2`: they are inconsistent; some are actually rings others are just slices, which leads to them behaving differently and the only way to know is to dig into the implementation. Some uses need to know the number of elements in the iterator before working through it. Wouldn't at that point a slice do as well?. I thought we needed iterators because a) we might not know in advance how many elements will be returned and b) collecting elements eagerly would be much more inefficient than using an iterator.

Especially regarding inefficiency of using slices over iterators: I now believe just behaving similar to `io.Reader` (but for Hashes instead of Bytes) would solve a lot of the complexity and remove the need for `rangesync.Seq`. Instead we could replace it with a `rangesync.Reader` that looks like this:

```go
type Reader interface {
    Read([]KeyBytes) (int, error) // what is the need for `KeyBytes` btw? isn't that always a types.Hash32?
}
```

The caller can still iterate over the elements (by reading in a loop) error propagation is more straight forward and `rangesync.SeqResult` is then probably not needed at all? The underlying can be a mock, a DB, some in memory structure, even a network connection - just like for the o.g. `io.Reader` 🙂  wdyt?

ivan4th commented:
Thanks a lot for review!

> Reviewing this PR took me quite some time because it builds on code that I'm not yet very familiar with. There are quite a few things that I noticed when going through the code:
> 
> We callbacks in places where they make code more complex (and inefficient) than it needs to be. Why are errors propagated using callbacks?

You mean the fetcher interaction (`mesh_data.go`)? If so, I'll provide details in the reply to the relevant comment.

> I'm also unsure about the DB abstraction that `syncv2` builds on. Why do we have to implement something like this ourselves? Are there no existing solutions that we can use? Is it even needed? It's another layer of abstraction for mostly simple queries as far as I can tell, it also overlaps with the existing `sql/builder` package that exists to solve mostly the same challenges and was newly implemented only a few months ago.

Passing just `TableName`, `IDColumn`, `TimestampColumn` and `Filter` via `SyncedTable` which are really all that `syncv2` needs is IMO much more clean than writing out 7 SQL queries at a higher level in the code, which would basically make SQL table based sync a rather leaky abstraction; especially given that later, some queries may need to be added/replaced. Using AST to generate SQL is just safer (see `CustomQuery` example below) and more readable than `fmt.Sprintf(...)` with multiple long strings sometimes spanning several lines.

The idea was that given that we planned to introduce [rqlite/sql](https://github.com/rqlite/sql) dependency in any case (it is to be used for migration code cleanup and clean schema comparsion), we could use it's SQLite AST representation for SQL query generation as well,
as the resulting code is much cleaner than one based on concatenating parts of SQL queries, in particular given operator precedence in SQL. Other places where I intended to use `sql/expr` are:
* Bloom filters (#6332) -- to be re-benchmarked and finalized; the PR description contains motivation for using these expressions
* Basically replace `sql/builder`. It's now kind of footgun in some aspects, e.g. if we ever happen to pass an expression containing `OR` as `CustomQuery` without adding `(...)` around it, we'll mess up results in a funny way due to operator precedence: https://github.com/spacemeshos/go-spacemesh/blob/9092fe701922855261d03dd0e54d4699b653445b/sql/builder/builder.go#L109 With AST based approach, this cannot happen.

The tests contain [generated SQL](https://github.com/spacemeshos/go-spacemesh/blob/9092fe701922855261d03dd0e54d4699b653445b/sync2/sqlstore/syncedtable_test.go#L51-L59) so it can also be seen clearly. Hand-written SQL is often error-prone and thus it probably doesn't harm to state the intent twice in different ways :)

Other than that, `sql/expr` is just a very thin wrapper around AST structures defined in [rqlite/sql](https://github.com/rqlite/sql) to make them a bit more readable. I wouldn't introduce it's use in `syncv2` if they were only to be used for `atxs` table, but we'll also need to sync malfeasant identities, and later quite likely other objects as well.

> I'm becoming unsure about the extensive use of iterators in `syncv2`: they are inconsistent; some are actually rings others are just slices, which leads to them behaving differently and the only way to know is to dig into the implementation. Some uses need to know the number of elements in the iterator before working through it. Wouldn't at that point a slice do as well?. I thought we needed iterators because a) we might not know in advance how many elements will be returned and b) collecting elements eagerly would be much more inefficient than using an iterator.

Re the need for iterators themselves, there are places in the code where the pattern of invoking `yield` callback makes code quite a bit easier, e.g. `FPTree` traversal. There are other places where we need to combine sequences, making use of `iter.Pull`: https://github.com/spacemeshos/go-spacemesh/blob/9092fe701922855261d03dd0e54d4699b653445b/sync2/rangesync/combine_seqs.go This basically means the need to retrieve data on item-by-item basis.
In this case, one of the sequences may refer to the database, while other may refer to an `FPTree` containing items received via the recent sync mechanism.

The need for cyclic iterators arises from the currently used range representation: https://github.com/spacemeshos/go-spacemesh/blob/develop/dev-docs/sync2-set-reconciliation.md#range-representation
I followed the original range sync paper in that. Right now I'm not 100% convinced that this is the right approach and maybe later we could experiment with simpler non-cyclic range representation for sync, but right now that would entail too much rework.
While we still have to deal with cyclic iterators, the code should never assume that iterator is not cyclic, and the places where this assumption is not followed (thx for noticing these), it should be fixed. We could also make iterators actually always cyclic for consistency, although I'm less sure this is strictly necessary.

Returning temporary slices instead of iterators, which can sometimes turn out to contain million of items, may cause substantial GC pressure and consequently unnecessarily high CPU usage, which is rather pronounced with the old sync implementation.

> 
> Especially regarding inefficiency of using slices over iterators: I now believe just behaving similar to `io.Reader` (but for Hashes instead of Bytes) would solve a lot of the complexity and remove the need for `rangesync.Seq`. Instead we could replace it with a `rangesync.Reader` that looks like this:
> 
> ```go
> type Reader interface {
>     Read([]KeyBytes) (int, error) // what is the need for `KeyBytes` btw? isn't that always a types.Hash32?
> }
> ```
> 
> The caller can still iterate over the elements (by reading in a loop) error propagation is more straight forward and `rangesync.SeqResult` is then probably not needed at all? The underlying can be a mock, a DB, some in memory structure, even a network connection - just like for the o.g. `io.Reader` 🙂 wdyt?

I'm afraid that will make e.g. `FPTree` traversal will become quite a bit more complicated, as well as aforementioned iterator combining code. It can of course wrap an `iter.Seq` under the hood but then I'm unsure whether it's worth the effort

ivan4th committed:
sync2: address comments

ivan4th committed:
sync2: make most Seqs/SeqResults non-cyclic

ivan4th committed:
sync2: address comments

ivan4th committed:
sync2: refactor ATXHandler.Commit()

ivan4th committed:
sync2: fixup

ivan4th commented:
Made the sequences non-cyclic in most cases, except for the case of `sync2/sqltore.IDStore` interface where it's somewhat less obvious and probably not strictly necessary

fasmat commented on sync2/atxs.go:
```diff
@@ -0,0 +1,338 @@
+			return err
+		case !errors.Is(err, &fetch.BatchError{}):
+			h.logger.Debug("failed to download ATXs", zap.Error(err))
+		}
```
Sorry, I had this wrong in my previous suggestion. `errors.Is` is used to check for value equality, not type equality. We probably want to do this here instead:

```suggestion
		someSucceeded, err := h.getAtxs(ctx, cs)
		batchErr := &fetch.BatchError{}
		switch {
		case err == nil:
		case errors.Is(err, context.Canceled):
			return err
		case !errors.As(err, batchErr):
			h.logger.Debug("failed to download ATXs", zap.Error(batchErr))
		}
```

The reason being that `errors.As` will check (and assign) `err` to `batchErr` if the type allows it. `errors.Is` does an equality check which by default will always fail unless `err == &fetch.BatchError{}`. If the type implements `interface{ Is(error) bool }` then `errors.Is` will use the `Is` method instead. `fetch.BatchError` has such a method that correctly checks if `errors.Is` would be `true` for any of the contained errors so that a caller can treat a batch error as if it was an error of a single item fetched.

ivan4th replied:
Fixed, but it had to be `case !errors.As(err, &batchErr):` b/c `BatchError` methods have pointer receiver

fasmat commented on sync2/interface.go:
```diff
@@ -0,0 +1,36 @@
+	Host() host.Host
+	Peers() *peers.Peers
+	RegisterPeerHash(peer p2p.Peer, hash types.Hash32)
+}
```
I think you can just inline `system.AtxFetcher`:

```suggestion
type Fetcher interface {
	GetAtxs(context.Context, []types.ATXID, ...GetAtxOpt) error
	Host() host.Host
	Peers() *peers.Peers
	RegisterPeerHash(peer p2p.Peer, hash types.Hash32)
}
```

Interfaces should be defined on the user side not in a `system` package.

fasmat replied:
With my other suggestions this can be further simplified to

```go
type Fetcher interface {
	GetAtxs(context.Context, []types.ATXID, ...system.GetAtxOpt) error
	RegisterPeerHashes(peer p2p.Peer, hash []types.Hash32)
}
```

ivan4th replied:
Fixed by merging your branch

fasmat commented on sync2/interface.go:
```diff
@@ -0,0 +1,36 @@
+	system.AtxFetcher
+	Host() host.Host
+	Peers() *peers.Peers
+	RegisterPeerHash(peer p2p.Peer, hash types.Hash32)
```
Also this method exists twice on the fetcher implementation:

```go
RegisterPeerHashes(peer p2p.Peer, hash []types.Hash32)
RegisterPeerHash(peer p2p.Peer, hash types.Hash32)
```

would it make sense to drop the second and just use the first instead?

ivan4th replied:
Fixed by merging your branch

fasmat commented on sync2/fptree/fptree.go:
```diff
@@ -363,6 +363,7 @@ func (ft *FPTree) traverseFrom(
 }
 
 // All returns all the items currently in the tree (including those in the IDStore).
+// The sequence in SeqResult is either empty or infinite.
```
My linter complains that there is a meaningless assertion in `sync2/fptree/nodepool_test.go`:

```go
require.Nil(t, nil, idx3)
```

ivan4th replied:
A leftover from some older code, removed

fasmat commented on syncer/interface.go:
```diff
@@ -47,7 +48,7 @@ type fetcher interface {
 	GetCert(context.Context, types.LayerID, types.BlockID, []p2p.Peer) (*types.Certificate, error)
 
-	system.AtxFetcher
+	sync2.Fetcher
```
This Interface can be simplified to:

```go
// fetcher is the interface to the low-level fetching.
type fetcher interface {
	GetLayerData(context.Context, p2p.Peer, types.LayerID) ([]byte, error)
	GetLayerOpinions(context.Context, p2p.Peer, types.LayerID) ([]byte, error)
	GetCert(context.Context, types.LayerID, types.BlockID, []p2p.Peer) (*types.Certificate, error)

	sync2.Fetcher

	GetMalfeasanceProofs(context.Context, []types.NodeID) error
	GetBallots(context.Context, []types.BallotID) error
	GetBlocks(context.Context, []types.BlockID) error
	RegisterPeerHashes(peer p2p.Peer, hashes []types.Hash32)

	SelectBestShuffled(int) []p2p.Peer
	PeerEpochInfo(context.Context, p2p.Peer, types.EpochID) (*fetch.EpochData, error)
	PeerMeshHashes(context.Context, p2p.Peer, *fetch.MeshHashRequest) (*fetch.MeshHashes, error)
}
```

fasmat replied:
With my other suggestions this can be further simplified to:

```go
type fetcher interface {
	GetLayerData(context.Context, p2p.Peer, types.LayerID) ([]byte, error)
	GetLayerOpinions(context.Context, p2p.Peer, types.LayerID) ([]byte, error)
	GetCert(context.Context, types.LayerID, types.BlockID, []p2p.Peer) (*types.Certificate, error)

	GetAtxs(context.Context, []types.ATXID, ...system.GetAtxOpt) error
	GetMalfeasanceProofs(context.Context, []types.NodeID) error
	GetBallots(context.Context, []types.BallotID) error
	GetBlocks(context.Context, []types.BlockID) error
	RegisterPeerHashes(peer p2p.Peer, hashes []types.Hash32)

	SelectBestShuffled(int) []p2p.Peer
	PeerEpochInfo(context.Context, p2p.Peer, types.EpochID) (*fetch.EpochData, error)
	PeerMeshHashes(context.Context, p2p.Peer, *fetch.MeshHashRequest) (*fetch.MeshHashes, error)
}
```

ivan4th replied:
Fixed by merging your branch

fasmat commented on syncer/syncer.go:
```diff
@@ -204,6 +246,17 @@ func NewSyncer(
+		serverOpts := append(
+			s.cfg.ReconcSync.ServerConfig.ToOpts(),
+			server.WithHardTimeout(s.cfg.ReconcSync.HardTimeout))
+		s.dispatcher = sync2.NewDispatcher(s.logger, fetcher.(sync2.Fetcher), serverOpts)
```
This cast shouldn't be necessary:

```suggestion
		s.dispatcher = sync2.NewDispatcher(s.logger, fetcher, serverOpts)
```

Also this here is the reason why the `sync2.Fetcher` interface needs to expose the `Host` method. I think instead it would be better to pass the host explicitly to `NewSyncer` and remove the method from the interface and `fetch.Fetch` implementation.

ivan4th replied:
Fixed by merging your branch

fasmat commented on sync2/atxs.go:
```diff
@@ -0,0 +1,338 @@
+	curSet := dbset.NewDBSet(db, atxsTable(epoch), 32, cfg.MaxDepth)
+	handler := NewATXHandler(logger, f, cfg.BatchSize, cfg.MaxAttempts,
+		cfg.MaxBatchRetries, cfg.FailedBatchDelay, nil)
+	return NewP2PHashSync(logger, d, name, curSet, 32, f.Peers(), handler, cfg, enableActiveSync)
```
This is the only use of the `fetcher.Peers()` method. Just like with `host` the `peer.Peers` service should be passed explicitly to `NewATXSyncer` and not indirectly via the fetcher. For this it needs to be initialised in `node.go` and passed to both the fetcher and here instead of constructing it in the fetcher.

ivan4th replied:
Fixed by merging your branch

fasmat reviewed (commented):
I added more comments to the current PR, sorry for taking a while to finish reviewing it but it is a rather large change...

Regarding your comment:

> Returning temporary slices instead of iterators, which can sometimes turn out to contain million of items, may cause substantial GC pressure and consequently unnecessarily high CPU usage, which is rather pronounced with the old sync implementation.

I do understand this, but using iterators doesn't necessarily solve the problem when what is iterated is a slice (which it is in many cases). Slices - as long as they are not copied - are just pointers that are returned or passed as argument. Wrapping them into an iterator doesn't necessarily mean there are fewer memory allocations or less pressure on the GC.

I do understand that especially with a DB backend it can be much more efficient to use iterators than slices. That's why I suggested a "`io.Reader` inspired" alternative. Were not all data has to be read into memory at once, but only as much as really is needed - similar how the iterators behave in that case.

> I'm afraid that will make e.g. FPTree traversal will become quite a bit more complicated, as well as aforementioned iterator combining code. It can of course wrap an iter.Seq under the hood but then I'm unsure whether it's worth the effort

Yes, it would still be possible to use `iter.Seq` where it would simplify code by just wrapping the interface in an iterator.

With the removal of passing along the length of the iterator and removing (most) cyclic cases it is probably OK now but I still feel like in many cases it neither improves readability nor performance but just adds a layer of abstraction 🤷

fasmat committed:
Simplify interfaces

ivan4th committed:
Fixup after Fetcher update

ivan4th committed:
Address comments

ivan4th commented:
* fixed fetcher/peer related stuff by merging @fasmat's branch
* hash fetcher logic refactoring is a must, but probably after we have mainnet switched to `syncv2` for both ATXs and malfeasance proofs, so that we have less code to refactor
  * also will need to remove non-streaming fetcher client/server
* other comments addressed

ivan4th commented:
bors merge
